reactivesocket-tck-drivers/src/main/java/io/reactivesocket/tckdrivers/common/TestSubscriber.java

```diff
@@ -0,0 +1,1115 @@
+                ByteBufferUtil.toUtf8String(p.getMetadata()));
+        System.out.println("ON NEXT GOT : "  + tup.getK() + " " + tup.getV());
+        if (isEcho) {
+            echosub.add(tup);
+            return;
+        }
+        if (!checkSubscriptionOnce) {
+            checkSubscriptionOnce = true;
+            if (subscription.get() == null) {
+                errors.add(new IllegalStateException("onSubscribe not called in proper order"));
+            }
+        }
+        lastThread = Thread.currentThread();
+
+        if (establishedFusionMode == QueueSubscription.ASYNC) {
+            while ((p = qs.poll()) != null) {
+                values.add(new Tuple<>(ByteBufferUtil.toUtf8String(p.getData()),
+                        ByteBufferUtil.toUtf8String(p.getMetadata())));
+            }
+            return;
+        }
+
+        values.add(tup);
+        numOnNext.countDown();
+        takeLatch.countDown();
+
+        if (t == null) {
+            errors.add(new NullPointerException("onNext received a null Subscription"));
+        }
+
+        actual.onNext(new PayloadImpl(tup.getK(), tup.getV()));
+    }
+
+    @Override
+    public void onError(Throwable t) {
+        if (!checkSubscriptionOnce) {
+            checkSubscriptionOnce = true;
+            if (subscription.get() == null) {
+                errors.add(new NullPointerException("onSubscribe not called in proper order"));
+            }
+        }
+        try {
+            lastThread = Thread.currentThread();
+            errors.add(t);
+
+            if (t == null) {
+                errors.add(new IllegalStateException("onError received a null Subscription"));
+            }
+
+            actual.onError(t);
+        } finally {
+            done.countDown();
+        }
+    }
+
+    @Override
+    public void onComplete() {
+        isComplete = true;
+        if (!checkSubscriptionOnce) {
+            checkSubscriptionOnce = true;
+            if (subscription.get() == null) {
+                errors.add(new IllegalStateException("onSubscribe not called in proper order"));
+            }
+        }
+        try {
+            lastThread = Thread.currentThread();
+            completions++;
+
+            actual.onComplete();
+        } finally {
+            done.countDown();
+        }
+    }
+
+    @Override
+    public final void request(long n) {
+        if (!SubscriptionHelper.validate(n)) {
+            return;
+        }
+        Subscription s = subscription.get();
+        if (s != null) {
+            s.request(n);
+        } else {
+            BackpressureHelper.add(missedRequested, n);
+            s = subscription.get();
+            if (s != null) {
+                long mr = missedRequested.getAndSet(0L);
+                if (mr != 0L) {
+                    s.request(mr);
+                }
+            }
+        }
+    }
+
+    public final void setEcho(EchoSubscription echosub) {
```

Comment:
This seems odd to need a one-off concrete implementation here. Why does TestSubscriber need to know about this?